README.md
hershey-fonts : Hershey vector fonts library, tools, and glyphs
===============================================================


Project home page: <http://www.whence.com/hershey-fonts/>
---------------------------------------------------------

  - Kamal Mostafa <kamal@whence.com>
  - Dr. A. V. Hershey
  - James Hurt


Package components
------------------

### libhersheyfont and hershey-font apps

Kamal Mostafa's **libhersheyfont** (a C library for loading .jhf format
Hershey font files) along with various utility programs such as
**hershey-font-gnuplot** (a Hershey text front-end for _gnuplot_).

### hershey-fonts-data

The Hershey fonts are a collection of vector fonts developed circa 1967
by Dr. A. V. Hershey.  Included are Latin, Greek, Cyrilic, Japanese, and
various symbol glyph sets encoded as .jhf format Hershey font files.
The .jhf font files were converted from Hershey's original NTIS format
files by James Hurt (see hershey-fonts.notes).
<http://en.wikipedia.org/wiki/Hershey_font>.


Build and install instructions
------------------------------

	cmake .
	make
	make test
	sudo make install     # optional

### The `HERSHEY_FONTS_DIR` environment variable

By default, **libhersheyfont** finds .jhf font files installed into the
system directory /usr/share/hershey-fonts/ or /usr/local/share/hershey-fonts/
depending on the build configuration.

To use **libhersheyfont** apps without installing, set the `HERSHEY_FONTS_DIR`
environment variable, e.g.:

	export HERSHEY_FONTS_DIR=$PWD/hershey-fonts
	tools/hershey-font-check futural
	tools/hershey-font-gnuplot . . .


License
-------

The **libhersheyfont** library and apps are
Copyright © 2013 Kamal Mostafa <kamal@whence.com>, licensed GPLv2+:
GNU GPL version 2 or later <http://gnu.org/licenses/gpl.html>.
The Hershey font glyph data is covered by a permissive use and redistribution
license.  This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.



debian/changelog
hershey-fonts (0.1-1) unstable; urgency=low

  * Initial release (Closes: #731014)

 -- Kamal Mostafa <kamal@whence.com>  Mon, 02 Dec 2013 13:54:15 -0800


tools/hershey-font-check.c
//
// hershey-font-check.c - Copyright (C) 2013 Kamal Mostafa <kamal@whence.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
//

#include <hersheyfont.h>
#include <stdio.h>
#include <string.h>


int
main( int argc, char **argv )
{
    if ( argc != 2 ) {
	fprintf(stderr,
		"usage: hershey-font-check {fontname | fontfile.jhf}\n");
	return 1;
    }

    // load a hershey_font structure either by filename or fontname
    const char *fontname = argv[1];
    struct hershey_font *hf;

    if ( strchr(fontname,'.') )
	hf = hershey_jhf_font_load(fontname);
    else
	hf = hershey_font_load(fontname);

    if ( !hf ) {
	perror(fontname);
	return 1;
    }

    // dump all the glyph data
    int c, i;
    for ( c=0; c<256; c++ ) {
	// get the hershey_glyph for ASCII character c
	struct hershey_glyph *hg = hershey_font_glyph(hf, c);

	// check whether there actually is a glyph for this character
	if ( hg->width == 0 )
	    continue;

	printf("  [[ %c ]] glyph(%d) width=%u npaths=%u\n",
			c, c, hg->width, hg->npaths);

	// walk the paths-list for this glyph
	struct hershey_path *hp;
	for ( hp=hg->paths; hp; hp=hp->next ) {
	    // begin draw path
	    printf("\t\tpath: nverts=%d\t", hp->nverts);
	    for ( i=0; i<hp->nverts; i++ ) {
		short x = hp->verts[i].x /* + x_render_pos */ ;
		short y = hp->verts[i].y;
		printf(" {%d,%d}", x, y);
	    }
	    // end draw path
	    printf("\n");
	}

	// note: to render a string of character glyphs left-to-right,
	// advance the x_render_pos by the width of each glyph here, e.g.:
	//
	// x_render_pos += hg->width;
    }

    // destroy the hershey_font
    hershey_font_free(hf);

    return 0;
}


libhersheyfont/hersheyfont.h
//
// hersheyfont.h - Copyright (C) 2013 Kamal Mostafa <kamal@whence.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
//

#ifndef HERSHEYFONT_H
#define HERSHEYFONT_H

struct hershey_vertex {
	short			x;
	short			y;
};

struct hershey_path {
	struct hershey_path	*next;
	unsigned short		nverts;
	struct hershey_vertex	verts[];
};

struct hershey_glyph {
	unsigned int		glyphnum;
	unsigned short		width;
	unsigned short		npaths;
	struct hershey_path	*paths;
};

struct hershey_font {
	struct hershey_glyph	glyphs[256];	/* indexed by ASCII value */
};


//
// Load Hershey fontname (e.g. "rowmans") into a new struct hershey_font.
// The associated .jhf font file will be loaded from .../share/hershey-fonts/.
//
struct hershey_font *	hershey_font_load( const char *fontname );


//
// Load a Hershey font .jhf file into a new struct hershey_font.
//
struct hershey_font *	hershey_jhf_font_load( const char *jhffile );

//
// Return a pointer to the struct hershey_glyph for ASCII character c.
//
struct hershey_glyph *	hershey_font_glyph( struct hershey_font *hf,
					    unsigned char c );

//
// Free a hershey_font allocated by hershey_font_load or hershey_jhf_font_load.
//
void			hershey_font_free( struct hershey_font *hf );

#endif // HERSHEYFONT_H


tools/hershey-font-gnuplot.c
//
// hershey-font-gnuplot.c - Copyright (C) 2013 Kamal Mostafa <kamal@whence.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
//

#include <hersheyfont.h>
#include <stdio.h>
#include <string.h>
#include <getopt.h>
#include <stdlib.h>

int sample_sheet_mode = 1;

char *render_text;

void
usage()
{
    fprintf(stderr,
"usage:  hershey-font-gnuplot [options] {font} 'Text to Render'\n"
"        hershey-font-gnuplot [options] {font}      - generate sample sheet\n"
"                {font}    a Hershey fontname or fontfile.jhf\n"
"                [options]\n"
"                   -T 'gnuplot_terminal_type [opts]'\n"
"                             e.g.:  -T 'pdf'\n"
"                                    -T 'png crop transparent'\n"
"                   -h font_height          (default 100)\n"
"examples:\n"
"    hershey-font-gnuplot futural   'Hello World!' | gnuplot -p\n"
"    hershey-font-gnuplot gothiceng 'Calligraphy'  | gnuplot -p\n"
"    hershey-font-gnuplot -T 'png crop' scriptc 'Nice Text' | gnuplot > out.png\n"
	);


}

int
main( int argc, char **argv )
{
    char *gnuplot_term_opts = "wxt";
    int gnuplot_height = 100;

    int opt;
    while ((opt = getopt(argc, argv, "h:T:")) != -1) {
	switch (opt) {
	case 'h':
	    gnuplot_height = strtoul(optarg, 0, 0);
	    break;
	case 'T':
	    gnuplot_term_opts = optarg;
	    break;
	default: /* '?' */
	   usage();
	   return 1;
	}
    }

    opt = argc - optind;
    if ( opt < 1 || opt > 2 ) {
	usage();
	return 1;
    }
    if ( opt == 2 )
	sample_sheet_mode = 0;


    // load a hershey_font structure either by filename or fontname
    const char *fontname = argv[optind++];
    struct hershey_font *hf;

    if ( strchr(fontname,'.') )
	hf = hershey_jhf_font_load(fontname);
    else
	hf = hershey_font_load(fontname);

    if ( !hf ) {
	perror(fontname);
	return 1;
    }

    // render the text

    printf("#\n# gnuplot input file generated by hershey-font-gnuplot\n#\n\n");

    printf("\n");

    int terminal_width, terminal_height;

    char render_text_buf[256];
    if ( sample_sheet_mode ) {
	int i;
	for ( i=0; i<256-32; i++ )
	    render_text_buf[i] = i+32;
	render_text = render_text_buf;
	terminal_width = 16 * gnuplot_height;
	terminal_height = 8 * gnuplot_height;
    } else {
	render_text = argv[optind++];
	terminal_width = 16 * gnuplot_height;
	terminal_height = gnuplot_height;
    }

    printf("set terminal %s size %d,%d\n",
	    gnuplot_term_opts, terminal_width, terminal_height);
    printf("\n");

    printf("unset xtics\n");
    printf("unset ytics\n");
    printf("unset border\n");
    printf("unset key\n");
    if ( sample_sheet_mode ) {
	printf("set title \"%s\"\n", fontname );
	printf("set size ratio 0.5\n");
	printf("set xrange [0:%d]\n", 32*32);
	printf("set yrange [0:%d]\n", 8*64);
    } else {
	printf("set size ratio 0.125\n");
	printf("set xrange [0:%d]\n", 16*16);
	printf("set yrange [0:%d]\n", 2*16);
    }
    printf("set style arrow 1 nohead\n" );
    printf("set style line 1 lc rgb \"black\" lw 0.5\n" );

    int x_render_pos = 0;
    int y_render_pos = 8;

    const char *p;
    for ( p=render_text; *p; p++ ) {

	// get the character c to be rendered
	int c = *((unsigned char *)p);

	// get the hershey_glyph for ASCII character c
	struct hershey_glyph *hg = hershey_font_glyph(hf, c);

	// check whether there actually is a glyph for this character
	if ( hg->width == 0 )
	    continue;

	// handle special placement for sample_sheet_mode
	if ( sample_sheet_mode ) {
	    x_render_pos = (c % 32) * 32;
	    y_render_pos = (c / 32) * 64;
	    y_render_pos = 8*64 - y_render_pos;
	}

	printf("#  [[ %c ]] glyph(%d) width=%u npaths=%u\n",
			c, c, hg->width, hg->npaths);

	// walk the paths-list for this glyph
	struct hershey_path *hp;
	short px, py;
	for ( hp=hg->paths; hp; hp=hp->next ) {
	    // begin draw path
	    printf("#\t\tpath: nverts=%d\n", hp->nverts);
	    int i;
	    for ( i=0; i<hp->nverts; i++ ) {
		short x = hp->verts[i].x + x_render_pos;
		short y = hp->verts[i].y + y_render_pos;
		if ( i > 0 )
		    printf("set arrow from %d,%d to %d,%d as 1 ls 1\n",
			    px, py, x, y);
		px = x;
		py = y;
	    }
	    // end draw path
	}

	// advance the x_render_pos by the width of this glyph
	x_render_pos += hg->width;
    }

    printf("plot NaN notitle\n");

    // destroy the hershey_font
    hershey_font_free(hf);

    return 0;
}


tools/hershey-jhf-fix-linebreaks.py
#!/usr/bin/python
#
# hershey-jhf-fix-linebreaks.py  -  fix linebreaks in .jhf Hershey font files
# Copyright (C) 2013 Kamal Mostafa <kamal@whence.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#

import sys

output = ''


def hershey_val(character):
    return ord(character) - ord('R')


def read_hershey_glyph(f):
    line = ''
    while ( line == '' ):
	line = f.readline()
	if (not line) or line == '\x1a':	# eof
	    return False
	line = line.rstrip()

    # read a Hershey format line
    glyphnum = int(line[0:5])		# glyphnum (junk in some .jhf files)
    nverts = int(line[5:8]) - 1
    leftpos = hershey_val(line[8])
    rightpos = hershey_val(line[9])
    vertchars = line[10:]
    nvertchars = len(vertchars)

    # join split lines in the Hershey data
    while ( nverts * 2 > nvertchars ):
	nextline = f.readline().rstrip()
	line += nextline
	vertchars += nextline
	nvertchars = len(vertchars)
    if ( nverts * 2 != nvertchars ):
	print >> sys.stderr, "hershey2olfont: Hershey format parse error (nvertchars=%d not %d)" % (nvertchars, nverts*2)
	sys.exit(1)

    # emit the fixed (joined) line
    global output
    output += "%s\n" % line

    return True


#
# main
#

if ( len(sys.argv) != 3 ):
    print >> sys.stderr, "usage: hershey-jhf-fix-linebreaks.py in.jhf out.jhf   (same filename is ok)"
    sys.exit(1)

hershey_in = sys.argv[1]
hershey_out = sys.argv[2]

with open(hershey_in) as f:
    while read_hershey_glyph(f):
	pass
    f.close()

fd = open(hershey_out,"w")
fd.write(output)
fd.close()



libhersheyfont/hershey-font-example.c
//
// hershey-font-example.c - Copyright (C) 2013 Kamal Mostafa <kamal@whence.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
//

#include <hersheyfont.h>
#include <stdio.h>


int
main( int argc, char **argv )
{
    const char *render_text = "HELLO";

    //
    // Load a Hershey font with hershey_font_load().
    //

    const char *fontname = "rowmans";
    struct hershey_font *hf = hershey_font_load(fontname);
    if ( !hf ) {
	perror(fontname);
	return 1;
    }

    //
    // Render each character's line paths with hershey_font_glyph().
    //

    int x_render_pos = 0;

    const char *p;
    for ( p=render_text; *p; p++ ) {

	// get the character c to be rendered
	int c = *p;

	// get the hershey_glyph for character c
	// and check whether there actually is a glyph for this character
	struct hershey_glyph *hg = hershey_font_glyph(hf, c);
	if ( hg->width == 0 )
	    continue;

	printf("# %c\n", c);

	// walk the paths-list for this glyph
	struct hershey_path *hp;
	for ( hp=hg->paths; hp; hp=hp->next ) {
	    // begin draw path
	    printf("\t");
	    int i;
	    for ( i=0; i<hp->nverts; i++ ) {
		short x = hp->verts[i].x + x_render_pos;
		short y = hp->verts[i].y;
		printf(" {%d,%d}", x, y);
	    }
	    // end draw path
	    printf("\n");
	}

	// advance the x_render_pos by the width of this glyph
	x_render_pos += hg->width;
    }

    //
    // Destroy the hershey_font with hershey_font_free().
    //
    hershey_font_free(hf);

    return 0;
}


libhersheyfont/hersheyfont.c
//
// hersheyfont.c - Copyright (C) 2013 Kamal Mostafa <kamal@whence.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
//


#include "hersheyfont.h"

#include <malloc.h>
#include <assert.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <stdio.h>

#ifndef HERSHEY_FONTS_DIR
// #define HERSHEY_FONTS_DIR "/usr/local/share/hershey-fonts"
#endif

#undef DEBUG


//
// Convert "Hershey-values" (offset by ASCII 'R') to integers
//
static inline short hershey_val(char c) { return (c - 'R'); }


//
// Load the (*hg) structure with data from the jhf_line.
//
static int
hershey_jhf_load_glyph( struct hershey_glyph *hg, char *jhf_line )
{
    char buf[8];
    char *end;

    unsigned int glyphnum, nverts;
    int leftpos, rightpos;
    char *vertchars;

    int len = strlen(jhf_line);
    if ( jhf_line[len-1] == '\n' ) {
	jhf_line[len-1] = 0;
	len--;
    }
    if ( len < 10 ) {
	fprintf(stderr, "%s: line length %d is too short\n", __FUNCTION__, len);
	errno = ERANGE;
	return 0;
    }

    errno = 0;

    strncpy(buf, jhf_line+0, 5);
    glyphnum = strtoul(buf, &end, 10);
    buf[5] = 0;
    if ( errno )
	return 0;

    strncpy(buf, jhf_line+5, 3);
    buf[3] = 0;
    nverts = strtoul(buf, &end, 10);
    if ( errno )
	return 0;

    len -= 8;

    if ( nverts*2 != len ) {
	fprintf(stderr, "%s: expected %d (not %d) coord bytes for nverts=%u\n", __FUNCTION__, nverts*2, len, nverts);
	errno = ERANGE;
	return 0;
    }

    leftpos = hershey_val(jhf_line[8]);
    rightpos = hershey_val(jhf_line[9]);
    if ( leftpos > rightpos ) {
	fprintf(stderr, "%s: bogus leftpos %d > rightpos %d\n", __FUNCTION__, leftpos, rightpos);
	errno = ERANGE;
	return 0;
    }

    // skip over the (leftpos,rightpos) pair
    nverts--;
    vertchars = jhf_line+10;

    // stuff the struct hershey_glyph

    hg->glyphnum = glyphnum;
    hg->width = rightpos - leftpos;	// FIXME add + 1 ??
    hg->npaths = 0;

    // split out the seperate line paths
    struct hershey_path *hp, *hplast=0;
    int i;
    int j = 0;
    for ( i=1; i<=nverts; i++ ) {
	// pen up
	if ( i==nverts || (vertchars[i*2]==' ' && vertchars[i*2+1]=='R') ) {

	    int npathverts = i - j;

	    unsigned long path_nbytes =
		    sizeof(struct hershey_path)
		    + npathverts * sizeof(struct hershey_vertex);

	    hp = malloc(path_nbytes);
#ifdef DEBUG
	    printf("... hp = malloc(%lu + %lu)\n",
		    sizeof(struct hershey_path),
		    npathverts * sizeof(struct hershey_vertex), hp );
#endif
	    assert(hp);

	    hp->next = 0;
	    hp->nverts = npathverts;
	    // Copy the vertices into hp->verts with y-invert and offsets to
	    // place the glyph origin at its lower left baseline corner with
	    // standard Cartesian coordinates.
	    int xoffset = -leftpos;	// shift left edge to 0
	    int yoffset = 16 - 7;	// (height - baseline) FIXME? hardcoded
	    int k;
	    for ( k=0; k<npathverts; k++ ) {
#ifdef DEBUG
		printf("... copy hp->verts[%d] = vertchars[%d*2]\n", k, k+j);
#endif
		hp->verts[k].x = xoffset + hershey_val(vertchars[(k+j)*2]);
		hp->verts[k].y = yoffset - hershey_val(vertchars[(k+j)*2 + 1]);
	    }

	    j = i + 1;

	    if ( hplast )
		hplast->next = hp;
	    else
		hg->paths = hp;
	    hg->npaths++;
	    hplast = hp;
	}
    }

    return 1;
}


//
// Load a Hershey font .jhf file into a new struct hershey_font.
//
struct hershey_font *
hershey_jhf_font_load( const char *jhffile )
{
    struct hershey_font *hf;

    FILE *fp = fopen(jhffile, "r");
    if (!fp)
	return 0;

    hf = calloc(1, sizeof(struct hershey_font));
    assert(hf);

#define BUFSIZE 4096

    char linebuf[BUFSIZE];
    int linecount = 0;
    int glyph_index = 32;

    while ( fgets(linebuf, BUFSIZE, fp) ) {
	int r;
	struct hershey_glyph *hg;
	hg = &(hf->glyphs[glyph_index++]);
	r = hershey_jhf_load_glyph(hg, linebuf);
	linecount++;
	if ( ! r ) {
	    perror(jhffile);
	    fprintf(stderr, "%s: ... at line number %d\n", jhffile, linecount);
	    hershey_font_free(hf);
	    hf = 0;
	    break;
	}
    }

    fclose(fp);

    return hf;
}


//
// Load Hershey fontname (e.g. "rowmans") into a new struct hershey_font.
// The associated .jhf font file will be loaded from .../share/hershey-fonts/.
//
struct hershey_font *
hershey_font_load( const char *fontname )
{
    const char *jhfdir;
    jhfdir = getenv("HERSHEY_FONTS_DIR");
    if ( !jhfdir )
	jhfdir = HERSHEY_FONTS_DIR;
    char *jhffile = malloc(strlen(jhfdir)+1+strlen(fontname)+5);
    assert(jhffile);
    sprintf(jhffile, "%s/%s.jhf", jhfdir, fontname);
    struct hershey_font *hf = hershey_jhf_font_load(jhffile);
    free(jhffile);
    return hf;
}


//
// Return a pointer to the struct hershey_glyph for ASCII character c.
//
struct hershey_glyph *
hershey_font_glyph( struct hershey_font *hf, unsigned char c )
{
    return &hf->glyphs[c];
}


//
// Free a hershey_font allocated by hershey_font_load or hershey_jhf_font_load.
//
void
hershey_font_free( struct hershey_font *hf )
{
    // free all the paths of each glyph
    int c;
    for ( c=0; c<256; c++ ) {
	const struct hershey_glyph *hg;
	struct hershey_path *hp, *next;
	hg = hershey_font_glyph(hf, c);
	for ( hp=hg->paths, next=hp; hp; hp=next ) {
	    next = hp->next;
	    free(hp);
	}
    }
    // free the hf structure itself
    free(hf);
}

